# Day 1 - Part C: Data preparation with Numpy and Pandas

Numpy and Pandas are python libraries which automate most of the data manipulation tasks for you. This notebook will provide an overview of these and guide you though data preparation steps including: cleaning, encoding and normalization.

Below is the code importing all the libraries that will be used throughout this notebook. Remember to run this line everytime you open the notebook. Otherwise, majority of the functions used here will not work!

In [1]:
import numpy as np
from numpy.linalg import inv

import pandas as pd
from sklearn.utils import resample
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold

# 1. Introduction to Numpy
Numpy adds support for large, multidimensional array manipulation. It supports operations such as addition, multiplication, inversion and many other. Here, you will learn how to create numpy arrays and matrices, and how to perform basic operations on them.

For more information refer to: https://docs.scipy.org/doc/numpy-dev/user/quickstart.html

## 1.1. Basic numpy operations
In numpy, you can create arrays from lists.

In [2]:
some_array = np.array([1, 2, 3, 4, 5])
print(some_array)

[1 2 3 4 5]


A single array can contain different objects: strings, floats, integers etc.

In [3]:
some_array = np.array(['blue', 'red', 'orange', 'white'])
print(some_array)

some_array = np.array([1.5, 2, 3.6, 4.1, 5.17])
print(some_array)

['blue' 'red' 'orange' 'white']
[ 1.5   2.    3.6   4.1   5.17]


You can access elements of an array using [ ], similarly as in python's lists. Indexes of elements start from 0, therefore the code below accesses the ** *2nd* ** element.

In [4]:
some_array[1]

2.0

Numpy arrays are mutable, therefore its elements can be changed by assigning a value. Below the ** *4th* ** element of an array has been changed to 15.

In [5]:
print("Before change:")
print(some_array)

some_array[3] = 15

print("\nAfter change:")
print(some_array)

Before change:
[ 1.5   2.    3.6   4.1   5.17]

After change:
[  1.5    2.     3.6   15.     5.17]


You can assign the whole numpy array to a variable, however be careful when you are doing it as modifying one array will modify the other array as well!

In [6]:
some_array = np.array([1.5, 2, 3.6, 4.1, 5.17])
some_array2 = some_array

print("Before change:")
print("some_array:", some_array)
print("some_array2:", some_array2)

some_array[3] = -6

print("\nAfter change:")
print("some_array:", some_array)
print("some_array2:", some_array2)

Before change:
some_array: [ 1.5   2.    3.6   4.1   5.17]
some_array2: [ 1.5   2.    3.6   4.1   5.17]

After change:
some_array: [ 1.5   2.    3.6  -6.    5.17]
some_array2: [ 1.5   2.    3.6  -6.    5.17]


As you see, elements in **both** arrays were modified even if only one of them was changed! It is because variables *some_array* and *some_array2* are pointers to arrays. We will not discuss pointers here, and you are not required to know how they work, however keep in mind that assigning numpy arrays to variables will "bound" them together and they will be susceptible to changes made on the other copy.

## 1.2 2D arrays
you can create arrays which are 2-dimensional.

In [7]:
np.array( [ [1, 2, 3], [3, 4, 5] ] )

array([[1, 2, 3],
       [3, 4, 5]])

To create a 2D array, you need to pass a list of columns to ** *np.array()* ** function, where each column is a list of row elements. You are provided with come examples below, but feel free to experiment with sizes!

Matrix 3x2:

In [8]:
np.array( [ [1, 2], [3, 3], [4, 5] ] )

array([[1, 2],
       [3, 3],
       [4, 5]])

Matrix 1x6:

In [9]:
np.array( [ [1, 2, 3, 3, 4, 5] ] )

array([[1, 2, 3, 3, 4, 5]])

Matrix 6x1:

In [10]:
np.array( [ [1], [2], [3], [3], [4], [5] ] )

array([[1],
       [2],
       [3],
       [3],
       [4],
       [5]])

## 1.3 Basic operations on arrays

Numpy is very efficient with array operations. Given two arrays A and B you can add them, multiply them, or calculate the inverse. Examples are provided to you below. For more information refer to: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matrix.html

In [11]:
A = np.array([ [1, 2], [0, 4]])
B = np.array([ [4, 0], [3, -1]])

print(A)
print()
print(B)

[[1 2]
 [0 4]]

[[ 4  0]
 [ 3 -1]]


In [12]:
A + B

array([[5, 2],
       [3, 3]])

In [13]:
A - B

array([[-3,  2],
       [-3,  5]])

In [14]:
A * B

array([[ 4,  0],
       [ 0, -4]])

To calculate inverse you have to import ** *inv* ** package. This is done for you in the first script at the beginning of this notebook.

In [15]:
inv(A)

array([[ 1.  , -0.5 ],
       [ 0.  ,  0.25]])

## 1.4 Functions for creating custom arrays
Numpy provides functions that enable to generate arrays for you. Examples are provided below.

** *np.zeros( (2, 2) )* ** creates 2x2 matrix containing only zeros, ** *np.ones( (2, 3) )* ** creates 2x3 matrix containing only ones, and ** *np.random.rand(3, 2)* ** creates 3x2 matrix containing random numbers between 0 - 1.

You can create 1-dimensional arrays, 2- or even more dimensional matrices.

In [16]:
np.zeros((2, 2))

array([[ 0.,  0.],
       [ 0.,  0.]])

In [17]:
np.ones((2, 3))

array([[ 1.,  1.,  1.],
       [ 1.,  1.,  1.]])

In [18]:
np.random.rand(3,2)

array([[ 0.77044942,  0.61046505],
       [ 0.82482842,  0.1298463 ],
       [ 0.46491163,  0.24932327]])

4D matrix 3x2x2x4:

In [19]:
np.random.rand(3, 2, 2, 4)

array([[[[ 0.46057643,  0.88457976,  0.15329158,  0.66469134],
         [ 0.63301954,  0.08287725,  0.06391142,  0.74054745]],

        [[ 0.76656553,  0.81986894,  0.79411521,  0.39608417],
         [ 0.23577641,  0.0554035 ,  0.51538962,  0.87632247]]],


       [[[ 0.97373072,  0.97620847,  0.76669715,  0.2283542 ],
         [ 0.01877245,  0.95357896,  0.89220995,  0.08922766]],

        [[ 0.88018606,  0.53191146,  0.46890286,  0.62010489],
         [ 0.9779534 ,  0.15542074,  0.11899297,  0.92844411]]],


       [[[ 0.62294831,  0.12010562,  0.2200909 ,  0.61108386],
         [ 0.99664771,  0.91706413,  0.79728883,  0.0046648 ]],

        [[ 0.79865085,  0.56806365,  0.33659093,  0.90927796],
         [ 0.58452455,  0.55686128,  0.47332551,  0.11053549]]]])

## 1.5 Slicing arrays
Numpy allows to slice the original array and work on the slide instead of the whole dataset

In [20]:
some_array = np.random.rand(3, 3)

print(some_array)

slice_array = some_array[0:2, 1:3]

print()
print(slice_array)

[[ 0.2989935   0.19927966  0.58283141]
 [ 0.71328775  0.16067147  0.65277094]
 [ 0.5315086   0.29988169  0.14839203]]

[[ 0.19927966  0.58283141]
 [ 0.16067147  0.65277094]]


You can specify the range which columns/rows should be included by using the format *array[ column_start:column_end, row_start:row_end ]*. Rows and columns in numpy (and also in Pandas) start indexing from 0, therefore if you want to include first and second column you need to use 0:2, if second and third use 1:3 etc. A notation 1:4 means including elements from index 1, 2 and 3. Note, that 4 is not included.

Again, remember that working on the slice will modify the original array too! The slice of an array has different indexing now, therefore the same element in the example below will have [0, 0] location in the slice array, and [0, 1] location in the original array.

In [21]:
print("Before change:")
print("some_array:\n", some_array)
print("some_array2:\n", slice_array)

slice_array[0, 0] = -6

print("\nAfter change:")
print("some_array:\n", some_array)
print("some_array2:\n", slice_array)

print()
print(slice_array[0, 0], some_array[0, 1])

Before change:
some_array:
 [[ 0.2989935   0.19927966  0.58283141]
 [ 0.71328775  0.16067147  0.65277094]
 [ 0.5315086   0.29988169  0.14839203]]
some_array2:
 [[ 0.19927966  0.58283141]
 [ 0.16067147  0.65277094]]

After change:
some_array:
 [[ 0.2989935  -6.          0.58283141]
 [ 0.71328775  0.16067147  0.65277094]
 [ 0.5315086   0.29988169  0.14839203]]
some_array2:
 [[-6.          0.58283141]
 [ 0.16067147  0.65277094]]

-6.0 -6.0


# 2. Introduction to Pandas

Pandas, similarly to numpy, provides support with large, multidimensional arrays. Additionally it is equipped with many more functions. The following section provides some key information about the use of the library for data manipulation. You are free to read more here: http://pandas.pydata.org/pandas-docs/stable/tutorials.html

## 2.1 Importing data from files
The ** *read_csv()* ** function enables to read CSV files, however pandas enables to read other formats e.g. excel files. Below, is an exemplary code snippet which reads the file *data.csv* and loads it into pandas DataFrame.

Useful arguments of the function:
- ** *sep* ** specifies the separator with which cells are separated with, 
- *** encoding* ** enables to change the file encoding. The default encoding is *utf-8*, however you might try *ISO-8859-1* if utf-8 does not work.

Run the code below to examine first rows of the dataset with ** *dataFrame.head()* ** function. Try to change the encoding to *utf-8* and see what happens!

In [22]:
dataFrame = pd.read_csv('./datasets/ecommerce_data/e_commerce.csv', sep=',', encoding='ISO-8859-1')
dataFrame.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 2.2 Creating DataFrame from a python dictionary
Alternatively, you can create Pandas DataFrame from a dictionay

In [23]:
data = [{'color_name': 'black', 'R': 0, 'G': 0, 'B': 0}, {'color_name': 'white', 'R': 255, 'G': 255, 'B': 255}, 
        {'color_name': 'red', 'R': 255, 'G': 0, 'B': 0}, {'color_name': 'blue', 'R': 0, 'G': 0, 'B': 255},
        {'color_name': 'green', 'R': 0, 'G': 255, 'B': 0}]

dataFrame = pd.DataFrame(data)
dataFrame

B    G    R color_name
0    0    0    0      black
1  255  255  255      white
2    0    0  255        red
3  255    0    0       blue
4    0  255    0      green

## 2.3 Inspecting the dataset

To inspect the dataset use ** *head()* ** function. This will show you first 5 rows of the dataset.

In [24]:
dataFrame.head()

B    G    R color_name
0    0    0    0      black
1  255  255  255      white
2    0    0  255        red
3  255    0    0       blue
4    0  255    0      green

Hoever, it will work only when ** *head()* ** is executed at the end of the code block. If you want to print your dataframe somewhere in the middle of the code block use:

In [25]:
print(dataFrame.head())

     B    G    R color_name
0    0    0    0      black
1  255  255  255      white
2    0    0  255        red
3  255    0    0       blue
4    0  255    0      green


## 2.4 Selecting, modifying and removing rows/columns

### Selecting a column
A good thing about Pandas is that you can refer to specific columns simply by using its name:

In [26]:
dataFrame['B']

0      0
1    255
2      0
3    255
4      0
Name: B, dtype: int64

In [27]:
dataFrame['color_name']

0    black
1    white
2      red
3     blue
4    green
Name: color_name, dtype: object

### Selecting rows
To select range of rows use the code below. DataFrame[2:4] selects all columns for rows with indexes 2 and 3 (4 is not included)

In [28]:
dataFrame[2:4]

B  G    R color_name
2    0  0  255        red
3  255  0    0       blue

If you want to select specific rows use ** *loc* **, or ** *iloc* **. loc will return the row that matches given label, whereas iloc will return a row that has ith position in the data frame.

In [29]:
dataFrame.loc[[0,2]]

B  G    R color_name
0  0  0    0      black
2  0  0  255        red

You can also select first e.g. 2 rows of the dataset, like below:

In [30]:
dataFrame[:2]

B    G    R color_name
0    0    0    0      black
1  255  255  255      white

Or last rows

In [31]:
dataFrame[-2:]

B    G  R color_name
3  255    0  0       blue
4    0  255  0      green

### Selecting multiple columns
You can select multiple columns by passing it as a list of column names, just like below:

In [32]:
column_slice = dataFrame[['B','color_name']]
column_slice.head()

B color_name
0    0      black
1  255      white
2    0        red
3  255       blue
4    0      green

While slicing this way, the slice created is not a pointer to the original data frame, therefore modifying it will not change the original dataset.

In [33]:
column_slice = dataFrame[['B','color_name']]

print("\nBefore changes:")
print(column_slice)
print()
print(dataFrame)

column_slice.loc[0, 'B'] = 14

print("\nAfter changes:")
print(column_slice)
print()
print(dataFrame)


Before changes:
     B color_name
0    0      black
1  255      white
2    0        red
3  255       blue
4    0      green

     B    G    R color_name
0    0    0    0      black
1  255  255  255      white
2    0    0  255        red
3  255    0    0       blue
4    0  255    0      green

After changes:
     B color_name
0   14      black
1  255      white
2    0        red
3  255       blue
4    0      green

     B    G    R color_name
0    0    0    0      black
1  255  255  255      white
2    0    0  255        red
3  255    0    0       blue
4    0  255    0      green


/Users/pascalwichmann/anaconda/envs/py36ML/lib/python3.6/site-packages/pandas/core/indexing.py:517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


You can find more on indexing here: http://pandas.pydata.org/pandas-docs/stable/indexing.html; and more on slice modification here: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy

### Column names and indices
You can return column names by using:

In [34]:
dataFrame.columns

Index(['B', 'G', 'R', 'color_name'], dtype='object')

You can see the row numbers/names by using:

In [35]:
dataFrame.index

RangeIndex(start=0, stop=5, step=1)

### Conditional selection
You can apply conditional selection by using ** dataFrame[*column_name*] *condition* *value* ** as specified below. The result of such expression is a True/False data frame which can be used to display rows which fulfil the condition. For example, the code below selects all rows that have *R* higher than 100.

In [36]:
select_greater_than_100 = (dataFrame['R'] > 100)
select_greater_than_100

0    False
1     True
2     True
3    False
4    False
Name: R, dtype: bool

In [37]:
dataFrame[select_greater_than_100]

B    G    R color_name
1  255  255  255      white
2    0    0  255        red

### Creating new columns
You can create a new column just by assigning values to non-existent column name. You can use completely new values, or create column using values from another one, as shown below: 

In [38]:
dataFrame['color_number'] = 0
dataFrame.head()

B    G    R color_name  color_number
0    0    0    0      black             0
1  255  255  255      white             0
2    0    0  255        red             0
3  255    0    0       blue             0
4    0  255    0      green             0

In [39]:
dataFrame['color_number'] = 65536 * dataFrame['R'] + 256 * dataFrame['G'] + dataFrame['B']
dataFrame.head()

B    G    R color_name  color_number
0    0    0    0      black             0
1  255  255  255      white      16777215
2    0    0  255        red      16711680
3  255    0    0       blue           255
4    0  255    0      green         65280

### Removing columns
You can remove the column by using ** *del* **:

In [40]:
del dataFrame['color_number']
dataFrame.head()

B    G    R color_name
0    0    0    0      black
1  255  255  255      white
2    0    0  255        red
3  255    0    0       blue
4    0  255    0      green

Or simply pop it from the dataset. This operation will remove the column and return its values creating new DataFrame

In [41]:
new_dataFrame = dataFrame.pop('color_name')
print(new_dataFrame)
print()
print(dataFrame)

0    black
1    white
2      red
3     blue
4    green
Name: color_name, dtype: object

     B    G    R
0    0    0    0
1  255  255  255
2    0    0  255
3  255    0    0
4    0  255    0


and now adding it back to the dataset

In [42]:
dataFrame['color_name'] = new_dataFrame
dataFrame.head()

B    G    R color_name
0    0    0    0      black
1  255  255  255      white
2    0    0  255        red
3  255    0    0       blue
4    0  255    0      green

# 3. The Data Science Process

The usual data science process contains the following steps:
1. Initial inspection of the dataset
2. Data preparation (cleaning, feature selection, normalization and splitting dataset into train/cross-validation/test)
4. Building the model
5. Evaluating the model performance

Before the machine learning can be applied to the dataset, a significant effort needs to be put into preparing the dataset for the model. Poorly performed data preparation might result in poor performance or unreliable results. This jupyter notebook will cover initial inspection of the dataset, data cleaning, normalisation and splitting dataset into train/cross-validation/test.

We will do so on the exemplary e-commerce dataset.

## 3.1 Data cleaning and initial inspection

In [43]:
dataFrame = pd.read_csv('./datasets/ecommerce_data/e_commerce.csv', sep=',', encoding='ISO-8859-1', dtype={
    'InvoiceNo': str, 
    'StockCode': str,
    'Description': str,
    'Quantity': np.float64,
    'InvoiceDate': str,
    'UnitPrice': np.float64,
    'CustomerID': np.float64,
    'Country': str
})
dataFrame.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER       6.0   
1    536365     71053                  WHITE METAL LANTERN       6.0   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER       8.0   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

We can see that the dataset includes 8 columns: *Invoice No.*, *StockCode*, *Description*, *Quantity*, *InvoiceDate*, *UnitPrice*, *CustomerID*, and *Country*.

Numerical variables include: *Quantity*, *UnitPrice*, and *CustomerID*.

Categorical variables include: *InvoiceNo*, *StockCode*, *Description*, and *Country*, as for now they are stored int the data frame as strings

There is only one datetime type variable: *InvoiceDate*, again stored in the data frame as a string.

You can inspect the types of the variables using:

In [44]:
dataFrame.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity       float64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

### 3.1.1 Replacing missing values
We will check if there are any variables that are missing.

In [45]:
dataFrame.isnull().any()

InvoiceNo      False
StockCode      False
Description     True
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID      True
Country        False
dtype: bool

We can see that there are *Description* and *CustomerID* features which are empty. The decision what to do with these depends on the circumstances. The most popular options are:

- remove missing values
- replace them with mean/most common category
- replace with the value of the previous row (forward-fill)
- replace with the value of the next row (backward-fill)

#### Removing missing values
Removing the missing values is done using ** *dropna()* ** function. It is important to assign reduced DataFrame to a variable, otherwise changes will not have any effect. Example of such behaviour is presented below.


In [46]:
print(dataFrame.shape)

dataFrame.dropna()

print(dataFrame.shape)

(238735, 8)
(238735, 8)


As you can see, the size of the DataFrame did not change, which means that the empty rows were **NOT** removed. Below is the corrected version of the code. The dataset has been reduced by almost 67 000 samples.

In [47]:
print(dataFrame.shape)

dataFrame = dataFrame.dropna()

print(dataFrame.shape)

(238735, 8)
(171810, 8)


In [48]:
dataFrame.isnull().any()

InvoiceNo      False
StockCode      False
Description    False
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID     False
Country        False
dtype: bool

#### Replacing missing values with the mean
First we will set *Quantity* in the row with index 1 to null, and then observe how it has been replaced.

In [49]:
dataFrame.loc[1, 'Quantity'] = np.nan
dataFrame.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER       6.0   
1    536365     71053                  WHITE METAL LANTERN       NaN   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER       8.0   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

Replacing missing values in column *Quantity* with mean value

In [50]:
dataFrame['Quantity'] = (dataFrame['Quantity'].replace(np.nan, dataFrame['Quantity'].mean()))
dataFrame.head()

InvoiceNo StockCode                          Description   Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER   6.000000   
1    536365     71053                  WHITE METAL LANTERN  12.254579   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER   8.000000   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE   6.000000   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.   6.000000   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

#### Replacing missing values with previous row (forward-fill)

First, setting null values in *StockCode* and *InvoiceDate* in the row with index 7. 

In [51]:
dataFrame.loc[7, 'StockCode'] = np.nan
dataFrame.loc[7, 'InvoiceDate'] = np.nan
dataFrame[3:10]

InvoiceNo StockCode                          Description  Quantity  \
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES       2.0   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER       6.0   
7    536366       NaN               HAND WARMER UNION JACK       6.0   
8    536366     22632            HAND WARMER RED POLKA DOT       6.0   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT      32.0   

      InvoiceDate  UnitPrice  CustomerID         Country  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7             NaN       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

Replacing missing values with the previous value (forward fill). Please note that this method will replace missing values in all columns, with values respective to that column; and might find very useful when replacing missing values in time-series.

In [52]:
dataFrame = dataFrame.fillna(method='ffill')
dataFrame[3:10]

InvoiceNo StockCode                          Description  Quantity  \
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES       2.0   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER       6.0   
7    536366     21730               HAND WARMER UNION JACK       6.0   
8    536366     22632            HAND WARMER RED POLKA DOT       6.0   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT      32.0   

      InvoiceDate  UnitPrice  CustomerID         Country  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7  12/1/2010 8:26       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

#### Replacing missing values with next row (backward-fill)

In [53]:
dataFrame.loc[7, 'StockCode'] = np.nan
dataFrame.loc[7, 'InvoiceDate'] = np.nan
dataFrame[3:10]

InvoiceNo StockCode                          Description  Quantity  \
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES       2.0   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER       6.0   
7    536366       NaN               HAND WARMER UNION JACK       6.0   
8    536366     22632            HAND WARMER RED POLKA DOT       6.0   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT      32.0   

      InvoiceDate  UnitPrice  CustomerID         Country  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7             NaN       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

In [54]:
dataFrame = dataFrame.fillna(method='bfill')
dataFrame[3:10]

InvoiceNo StockCode                          Description  Quantity  \
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES       2.0   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER       6.0   
7    536366     22632               HAND WARMER UNION JACK       6.0   
8    536366     22632            HAND WARMER RED POLKA DOT       6.0   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT      32.0   

      InvoiceDate  UnitPrice  CustomerID         Country  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7  12/1/2010 8:28       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

### 3.1.2 Initial inspection and data cleaning

Now we can investigate how the numerical dataset looks like by using ** *describe()* ** function.

In [55]:
dataFrame.describe()

Quantity      UnitPrice     CustomerID
count  171810.000000  171810.000000  171810.000000
mean       12.254579       3.798716   15274.144730
std       258.624198     101.355052    1724.536126
min    -74215.000000       0.000000   12346.000000
25%         2.000000       1.250000   13838.000000
50%         5.000000       1.950000   15130.000000
75%        12.000000       3.750000   16814.000000
max     74215.000000   38970.000000   18287.000000

It can be observed that the *Quantity* feature can have negative values, which is not what we would expect. These values can be removed from the dataset by applying conditional selection discussed before.

First, we see how many of these have non-positive *Quantity* values.

In [56]:
len(dataFrame[dataFrame['Quantity'] <= 0])

4098

Now, we will be selecting these rows using conditional selection.

In [57]:
print(dataFrame.shape)

indexes_with_negative_height = dataFrame[dataFrame['Quantity'] < 0].index
dataFrame = dataFrame.drop(indexes_with_negative_height)

print(dataFrame.shape)

(171810, 8)
(167712, 8)


We can observe that the dataset has been reduces exactly by 4098 rows.

Interestingly, there are 13 rows which have unit price equal to 0. This raises a warning flag since it is not often for an item to be for free. We will be displaying these rows.

In [58]:
len(dataFrame[dataFrame['UnitPrice'] == 0])

13

In [59]:
dataFrame[dataFrame['UnitPrice'] == 0]

InvoiceNo StockCode                          Description  Quantity  \
9302      537197     22841         ROUND CAKE TIN VINTAGE GREEN       1.0   
33576     539263     22580         ADVENT CALENDAR GINGHAM SACK       4.0   
40089     539722     22423             REGENCY CAKESTAND 3 TIER      10.0   
47068     540372     22090              PAPER BUNTING RETROSPOT      24.0   
47070     540372     22553               PLASTERS IN TIN SKULLS      24.0   
56674     541109     22168        ORGANISER WOOD ANTIQUE WHITE        1.0   
86789     543599    84535B         FAIRY CAKES NOTEBOOK A6 SIZE      16.0   
130188    547417     22062  CERAMIC BOWL WITH LOVE HEART DESIGN      36.0   
139453    548318     22055   MINI CAKE STAND  HANGING STRAWBERY       5.0   
145208    548871     22162          HEART GARLAND RUSTIC PADDED       2.0   
157042    550188     22636   CHILDS BREAKFAST SET CIRCUS PARADE       1.0   
187613    553000     47566                        PARTY BUNTING       4.0   
198383    554037     22619            SET OF 6 SOLDIER SKITTLES      80.0   

             InvoiceDate  UnitPrice  CustomerID         Country  
9302     12/5/2010 14:02        0.0     12647.0         Germany  
33576   12/16/2010 14:36        0.0     16560.0  United Kingdom  
40089   12/21/2010 13:45        0.0     14911.0            EIRE  
47068     1/6/2011 16:41        0.0     13081.0  United Kingdom  
47070     1/6/2011 16:41        0.0     13081.0  United Kingdom  
56674    1/13/2011 15:10        0.0     15107.0  United Kingdom  
86789    2/10/2011 13:08        0.0     17560.0  United Kingdom  
130188   3/23/2011 10:25        0.0     13239.0  United Kingdom  
139453   3/30/2011 12:45        0.0     13113.0  United Kingdom  
145208    4/4/2011 14:42        0.0     14410.0  United Kingdom  
157042   4/14/2011 18:57        0.0     12457.0     Switzerland  
187613   5/12/2011 15:21        0.0     17667.0  United Kingdom  
198383   5/20/2011 14:13        0.0     12415.0       Australia

However unusual, these items might still be valid entries because they might have been sold as an addition to other item, or bought with a voucher, therefore they will be kept in the dataset.

You can further inspect the dataset by calling  functions e.g. min(), max(), mean(), etc. explicitly on the dataFrame

In [60]:
dataFrame.max()

InvoiceNo                                557955
StockCode                                  POST
Description    ZINC WIRE SWEETHEART LETTER TRAY
Quantity                                  74215
InvoiceDate                       6/9/2011 9:54
UnitPrice                               8142.75
CustomerID                                18287
Country                             Unspecified
dtype: object

Note that for the strings the minimum and maximum values are evaluated based on the characters included in the string. For example, a letter "A" will be considered as "smaller" than letter "Z". Moreoever, string digits e.g. "8" are always considered to be smaller than letters.

In [61]:
dataFrame.min()

InvoiceNo                              536365
StockCode                               10002
Description     4 PURPLE FLOCK DINNER CANDLES
Quantity                                    1
InvoiceDate                   1/10/2011 10:32
UnitPrice                                   0
CustomerID                              12346
Country                             Australia
dtype: object

Similarly, you can apply functions ** *mode()* ** or ** *mean()* **.

More functions can be found here: https://pandas.pydata.org/pandas-docs/stable/api.html#api-dataframe-stats

You can check out the correlation coefficients between columns of numerical values by using *corr()* function. In the next notebook on data visualisation you will learn how to create a correlation heatmap.

In [62]:
dataFrame.corr()

Quantity  UnitPrice  CustomerID
Quantity    1.000000  -0.004563   -0.010708
UnitPrice  -0.004563   1.000000   -0.007881
CustomerID -0.010708  -0.007881    1.000000

### Invalid categorical values

If you want to inspect if the categorical column contains invalid entries you can use *value_counts()* function, as below.

In [63]:
dataFrame['Country'].value_counts()

United Kingdom          149183
Germany                   3949
France                    3531
EIRE                      2609
Netherlands               1139
Spain                     1131
Belgium                    914
Switzerland                679
Australia                  630
Portugal                   610
Norway                     368
Channel Islands            363
Cyprus                     350
Finland                    307
Italy                      299
Japan                      230
Sweden                     195
Denmark                    184
Poland                     180
Austria                    124
Singapore                  113
Iceland                    102
Greece                      85
Unspecified                 72
Canada                      68
Lebanon                     45
Malta                       45
Lithuania                   35
Brazil                      32
European Community          31
United Arab Emirates        30
USA                         22
Bahrain 

The same applies for numerical values

Most of the values seem to be valid entries, however some of the countries are categorized as "Unspecified". It is our decision whether we want to remove rows which have this value or not, and in this case we will do this be executing the following code:

In [64]:
print(dataFrame.shape)

dataFrame = dataFrame.drop(dataFrame[dataFrame['Country'] == 'Unspecified'].index)

print(dataFrame.shape)

(167712, 8)
(167640, 8)


where *dataFrame[dataFrame['Country'] == 'Unspecified']* selects all the rows that have 'Unspecified' country; ** *.index* ** returns indexes associated with these rows; and ** *dataFrame.drop()* ** removes these indexes from the dataset.

We have removed exactly 72 rows, as specified for 'Unspecified' by value_counts().

## 3.2 Encoding categorical variables

Most machine learning methods do not directly work with categories, there is a need to transform these into numerical values. There are two options to do it:

- encoding each category by number e.g. Monday - 1, Tuesday - 2 etc.
- creating dummy columns which encode the category. For example for day of the week, we would create 7 columns, each corresponsing to one day of the week.

To demonstrate both approaches, we will switch to the color dataset used at the beginning of the notebook.

In [65]:
data = [{'color_name': 'black', 'R': 0, 'G': 0, 'B': 0}, {'color_name': 'white', 'R': 255, 'G': 255, 'B': 255}, 
        {'color_name': 'red', 'R': 255, 'G': 0, 'B': 0}, {'color_name': 'blue', 'R': 0, 'G': 0, 'B': 255},
        {'color_name': 'green', 'R': 0, 'G': 255, 'B': 0}, {'color_name': 'green', 'R': 0, 'G': 255, 'B': 0}]

new_dataFrame = pd.DataFrame(data)
new_dataFrame['color_name2'] = new_dataFrame['color_name']
new_dataFrame

B    G    R color_name color_name2
0    0    0    0      black       black
1  255  255  255      white       white
2    0    0  255        red         red
3  255    0    0       blue        blue
4    0  255    0      green       green
5    0  255    0      green       green

### 3.2.1 Numerical encoding of categorical variables

First, will transform the type of the column to *category*, so that pandas map each category with the unique code.

In [66]:
new_dataFrame['color_name'] = new_dataFrame['color_name'].astype('category')

Next, we will update the column with corresponding category codes

In [67]:
new_dataFrame['color_name'] = new_dataFrame['color_name'].cat.codes
new_dataFrame

B    G    R  color_name color_name2
0    0    0    0           0       black
1  255  255  255           4       white
2    0    0  255           3         red
3  255    0    0           1        blue
4    0  255    0           2       green
5    0  255    0           2       green

We can see that each color has now each category assigned: *black* - 0, *white* - 4, *red* - 3, etc.

### 3.2.2 Creating dummy columns (also called as One Hot Encoding)

Execute ** *pd.get_dummies()* ** function to expand a specific column to dummy columns. *columns* parameter specifies the names of columns that will be expanded using one hot encoding.

In [68]:
new_dataFrame = pd.get_dummies(new_dataFrame, columns=['color_name2'])

In [69]:
new_dataFrame

B    G    R  color_name  color_name2_black  color_name2_blue  \
0    0    0    0           0                  1                 0   
1  255  255  255           4                  0                 0   
2    0    0  255           3                  0                 0   
3  255    0    0           1                  0                 1   
4    0  255    0           2                  0                 0   
5    0  255    0           2                  0                 0   

   color_name2_green  color_name2_red  color_name2_white  
0                  0                0                  0  
1                  0                0                  1  
2                  0                1                  0  
3                  0                0                  0  
4                  1                0                  0  
5                  1                0                  0

Now we can see that the *color_name2* column has been expanded to 5 binary columns.

## 3.3 Unbalanced datasets

Datasets often might contain unbalanced categorical variables; that means that some of these values appear more frequenty than the other. When the unbalanced feature is the target i.e. the feature that we want to predict using machine learning models, it is a problem because our model will be trained to predict the majority category for most of the time. More on this will be presented in the notebook dealing with evaluating performance of the machine learning model.

There are many different methods to deal with the problem of unbalanced datasets, however in this notebook we will focus on two of them: under-sampling and over-sampling. Under-sampling reduce the size of our original dataset removing instances of the majority class to match the minority class. Over-sampling adds new data points which are similar to these of the minority class. 

Please note, that under-sampling and over-sampling might not be the best to deal with unbalanced datasets, as it will introduce bias. For example, under-sampling will remove instances of the majority class that might be meaningful. However, it still might work in practice for some cases.

Under-sampling and over-sampling will be presented below, however before we do it, we will slighlty modify the e-commerce dataset.

Let's assume that we are interested only in transactions which happened in United Kingdom and France. Below is the code which will select all the rows which have 'United Kingdom' value in column *Country*, and these with 'France' value in the same column.

Both selections are then concatenated togehter using ** *pd.concat()* ** pandas function. Concatenate adds one data frame below the other; this implies that both data frames need to have the same columns. More about concatenation (and other merging functions) can be found here: https://pandas.pydata.org/pandas-docs/stable/merging.html

In [70]:
UKData = dataFrame.loc[dataFrame['Country'] == 'United Kingdom']
FranceData = dataFrame.loc[dataFrame['Country'] == 'France']

dataFrame = pd.concat([UKData, FranceData])
dataFrame

InvoiceNo StockCode                          Description    Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER    6.000000   
1         536365     71053                  WHITE METAL LANTERN   12.254579   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER    8.000000   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE    6.000000   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.    6.000000   
5         536365     22752         SET 7 BABUSHKA NESTING BOXES    2.000000   
6         536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER    6.000000   
7         536366     22632               HAND WARMER UNION JACK    6.000000   
8         536366     22632            HAND WARMER RED POLKA DOT    6.000000   
9         536367     84879        ASSORTED COLOUR BIRD ORNAMENT   32.000000   
10        536367     22745           POPPY'S PLAYHOUSE BEDROOM     6.000000   
11        536367     22748            POPPY'S PLAYHOUSE KITCHEN    6.000000   
12        536367     22749    FELTCRAFT PRINCESS CHARLOTTE DOLL    8.000000   
13        536367     22310              IVORY KNITTED MUG COSY     6.000000   
14        536367     84969   BOX OF 6 ASSORTED COLOUR TEASPOONS    6.000000   
15        536367     22623        BOX OF VINTAGE JIGSAW BLOCKS     3.000000   
16        536367     22622       BOX OF VINTAGE ALPHABET BLOCKS    2.000000   
17        536367     21754             HOME BUILDING BLOCK WORD    3.000000   
18        536367     21755             LOVE BUILDING BLOCK WORD    3.000000   
19        536367     21777          RECIPE BOX WITH METAL HEART    4.000000   
20        536367     48187                  DOORMAT NEW ENGLAND    4.000000   
21        536368     22960             JAM MAKING SET WITH JARS    6.000000   
22        536368     22913          RED COAT RACK PARIS FASHION    3.000000   
23        536368     22912       YELLOW COAT RACK PARIS FASHION    3.000000   
24        536368     22914         BLUE COAT RACK PARIS FASHION    3.000000   
25        536369     21756             BATH BUILDING BLOCK WORD    3.000000   
46        536371     22086      PAPER CHAIN KIT 50'S CHRISTMAS    80.000000   
47        536372     22632            HAND WARMER RED POLKA DOT    6.000000   
48        536372     22633               HAND WARMER UNION JACK    6.000000   
49        536373    85123A   WHITE HANGING HEART T-LIGHT HOLDER    6.000000   
...          ...       ...                                  ...         ...   
235974    557669     22629                  SPACEBOY LUNCH BOX    12.000000   
235975    557669     22631             CIRCUS PARADE LUNCH BOX    12.000000   
235976    557669     22319    HAIRCLIPS FORTIES FABRIC ASSORTED   12.000000   
235977    557669     23076           ICE CREAM SUNDAE LIP GLOSS   24.000000   
235978    557669     22328   ROUND SNACK BOXES SET OF 4 FRUITS     6.000000   
235979    557669     22326  ROUND SNACK BOXES SET OF4 WOODLAND     6.000000   
235980    557669     21881                       CUTE CATS TAPE   12.000000   
235981    557669     21883                     STARS GIFT TAPE    12.000000   
235982    557669     16237                 SLEEPING CAT ERASERS   30.000000   
235983    557669     22492              MINI PAINT SET VINTAGE    36.000000   
235984    557669     22727            ALARM CLOCK BAKELIKE RED     4.000000   
235985    557669     22726           ALARM CLOCK BAKELIKE GREEN    4.000000   
235986    557669     22728            ALARM CLOCK BAKELIKE PINK    4.000000   
235987    557669     22398           MAGNETS PACK OF 4 SWALLOWS   12.000000   
235988    557669     23158          SET OF 5 LUCKY CAT MAGNETS    12.000000   
235989    557669      POST                              POSTAGE    2.000000   
236116    557734     20974          12 PENCILS SMALL TUBE SKULL   48.000000   
236117    557734     23190  BUNDLE OF 3 SCHOOL EXERCISE BOOKS     12.000000   
236118    557734     22149           FELTCRAFT 6 FLOWER F

Now we can confirm that our dataset contain only entries related to United Kingdom and France. Moreover, we can observe that there are many more entries related to UK than to France. This means that this dataset is *unbalanced*, therefore we will be applying under-sampling and over-sampling.

In [71]:
dataFrame['Country'].value_counts()

United Kingdom    149183
France              3531
Name: Country, dtype: int64

### 3.3.1 Under-sampling

Under-sampling is removing the instances of the majority class so that it matches the number of minority class instances. We will do it using sklearn library ** *resample* **. First, we will create slices of the original dataset, where the first slice contains the majority class ("United Kingdom"), and the second slice contains the minorty class ("France").

In [72]:
df_majority = dataFrame[dataFrame['Country']=='United Kingdom']
df_minority = dataFrame[dataFrame['Country']=='France']
minority_len = (len(df_minority))

Now, we will run the resample method with parameter *replace* equal to False, and n_samples equal to the number of instances in the minority class. The reduced majority class will be concatenated with the minority class, and we can observe that there is now exactly the same number of instances for majority and minority classes.

In [73]:
df_majority_downsampled = resample(df_majority,replace=False,n_samples=minority_len,random_state=123)
ddown = pd.concat([df_minority, df_majority_downsampled])
ddown['Country'].value_counts()

France            3531
United Kingdom    3531
Name: Country, dtype: int64

### 3.3.2 Over-sampling

Over-sampling will create copies of dataset instances to match required number of samples. Below, we count number of instances in the majority class, and create additional instances in the minority class to match the majority class.

We can now see that there is the same number of instances for both classes, and it is equal to the initial size of the majority class.

In [74]:
majority_len = (len(df_majority))

df_minority_upsampled = resample(df_minority,replace=True,n_samples=majority_len,random_state=123)
dup = pd.concat([df_majority, df_minority_upsampled])
dup['Country'].value_counts()

France            149183
United Kingdom    149183
Name: Country, dtype: int64

## 3.4 Normalization

Most of the machine learning algorithms require the dataset they are trained on to be normalized. Below is the example of normalization of the e-commerce dataset.

First, we will be encoding categorical variables. We will use category codes for *InvoiceNo*, *StockCode* and *Description*, and One Hot Encoding for *Contry*.

In [75]:
dataFrame.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [76]:
dataFrame['StockCode'] = dataFrame['StockCode'].astype('category')
dataFrame['StockCode'] = dataFrame['StockCode'].cat.codes

dataFrame['Description'] = dataFrame['Description'].astype('category')
dataFrame['Description'] = dataFrame['Description'].cat.codes

dataFrame['InvoiceNo'] = dataFrame['InvoiceNo'].astype('category')
dataFrame['InvoiceNo'] = dataFrame['InvoiceNo'].cat.codes

dataFrame = pd.get_dummies(dataFrame, columns=['Country'])
dataFrame.head()

InvoiceNo  StockCode  Description   Quantity     InvoiceDate  UnitPrice  \
0          0       2886         3173   6.000000  12/1/2010 8:26       2.55   
1          0       2329         3181  12.254579  12/1/2010 8:26       3.39   
2          0       2514          774   8.000000  12/1/2010 8:26       2.75   
3          0       2468         1588   6.000000  12/1/2010 8:26       3.39   
4          0       2467         2416   6.000000  12/1/2010 8:26       3.39   

   CustomerID  Country_France  Country_United Kingdom  
0     17850.0               0                       1  
1     17850.0               0                       1  
2     17850.0               0                       1  
3     17850.0               0                       1  
4     17850.0               0                       1

The *InvoiceDate* field would need to be translated into the timestamp in order to be correcrtly normalized, however as it is not the scope of this notebook we will remove it from our data frame. If you are interested, you can read more about timestamps here: https://pandas.pydata.org/pandas-docs/stable/timeseries.html

In [77]:
del dataFrame['InvoiceDate']

We will use ** *preprocessing* ** package from sklearn to normalize our dataset. We have to normalize column by column. Otherwise the normalization will take place across different column.

The preprocessing package requires to pass the data in a form of a vector, therefore we will be first reshaping pandas data frame to a vector, and then reshaping it again as a column and updating the data frame.

In [78]:
for column in dataFrame.columns:
    vector = dataFrame[column].values.reshape(1, len(dataFrame[column]))
    normalized_vector = preprocessing.normalize(vector, norm="l2")
    dataFrame[column] = normalized_vector.reshape(len(dataFrame[column]), 1)
dataFrame.head()

/Users/pascalwichmann/anaconda/envs/py36ML/lib/python3.6/site-packages/sklearn/utils/validation.py:429: DataConversionWarning: Data with input dtype int16 was converted to float64 by the normalize function.
  warnings.warn(msg, _DataConversionWarning)
/Users/pascalwichmann/anaconda/envs/py36ML/lib/python3.6/site-packages/sklearn/utils/validation.py:429: DataConversionWarning: Data with input dtype uint8 was converted to float64 by the normalize function.
  warnings.warn(msg, _DataConversionWarning)


InvoiceNo  StockCode  Description  Quantity  UnitPrice  CustomerID  \
0        0.0   0.004574     0.004132  0.000078   0.000290    0.002934   
1        0.0   0.003691     0.004142  0.000160   0.000385    0.002934   
2        0.0   0.003984     0.001008  0.000105   0.000312    0.002934   
3        0.0   0.003911     0.002068  0.000078   0.000385    0.002934   
4        0.0   0.003910     0.003146  0.000078   0.000385    0.002934   

   Country_France  Country_United Kingdom  
0             0.0                0.002589  
1             0.0                0.002589  
2             0.0                0.002589  
3             0.0                0.002589  
4             0.0                0.002589

## 3.5 Splitting the dataset

In this section we will discuss two ways of splitting the datasets:
- Splitting inot train, cross-validation and test sets
- K-fold cross-validation

We will perform the split on the e-commerce dataset, where first we will extract the variable we want to predict - let it be *UnitPrice* in this case.

In [79]:
target = dataFrame.pop('UnitPrice')

### 3.5.1 Splitting into train, cross-validation, and test sets

To split the original dataset into three sets: train, cross-validation and test we can use ** *train_test_split()* ** function twice. First split will divide our dataset into (train + cross-validation) and the test set. The second split will divide the data into train and cross-validation. If we want to divide into 60% training, 20% cross-validation, and 20% test we should use at first 0.2 test size and 0.8 train size. While dividing for the second time we have to take into account that the dataset is now reduced, therefore we have to adjust the ratios to: 0.2/0.8 = 0.25 test size and 0.6/0.8 = 0.75 train size, which will give us train and cross-validation sets.

In [80]:
# split into (train + cross validation) and test sets
X_x, X_test, y_x, y_test = train_test_split(dataFrame, target, test_size=0.2, train_size=0.8, random_state=43)

# split into train and cross validation
X_train, X_cv, y_train, y_cv = train_test_split(X_x, y_x, test_size=0.25, train_size=0.75, random_state=43)

print(X_train.shape, X_cv.shape, X_test.shape)

(91628, 7) (30543, 7) (30543, 7)


### 3.5.2 K-fold

To perform K-fold you can use the class provided by scikit-learn ** *Kfold()* **. The argument ** *n_splits* ** defines how many parts we want to divide the dataset to, ** *shuffle* ** indicates whether we want to shuffle the dataset before we split it. Kfold will then return the indexes of the chosen rows, from which we can build our train and test subsets.

In [81]:
kf = KFold(n_splits=4, shuffle=True, random_state=43)

print("Original dataset:")
print(dataFrame.shape)

kfold_count = 0
for train_indexes, test_indexes in kf.split(dataFrame):
    train_dataset = pd.DataFrame(dataFrame.iloc[train_indexes])
    test_dataset = pd.DataFrame(dataFrame.iloc[test_indexes])
    train_target = pd.DataFrame(target.iloc[train_indexes])
    test_target = pd.DataFrame(target.iloc[test_indexes])
    
    kfold_count += 1
    print("Kfold", str(kfold_count) + "-iteration:")  
    
    print("Train:", train_dataset.shape, "Test:", test_dataset.shape)

Original dataset:
(152714, 7)
Kfold 1-iteration:
Train: (114535, 7) Test: (38179, 7)
Kfold 2-iteration:
Train: (114535, 7) Test: (38179, 7)
Kfold 3-iteration:
Train: (114536, 7) Test: (38178, 7)
Kfold 4-iteration:
Train: (114536, 7) Test: (38178, 7)


If you want to read more on splitting the dataset go to: http://scikit-learn.org/stable/modules/cross_validation.html